### PR TITLE
Fix stackoverflow with redefining resource

### LINKED ISSUE #8677: Stack overflow exception when loading StaticResource with same ResourceKey
**Describe the bug**

Stack overflow exception when loading StaticResource with same ResourceKey

**To Reproduce**
`
<StaticResource x:Key="SystemAccentColor" ResourceKey="SystemAccentColor" />
`

**Expected behavior**

**Screenshots**

tg_image_459510962

**Desktop (please complete the following information):**

- OS: [e.g. Windows, Mac, Linux (State distribution)]
- Version master after https://github.com/AvaloniaUI/Avalonia/pull/8422 merged

**Additional context**

### LINKED ISSUE #8831: Stack overflow when extending an existing ControlTheme
**Describe the bug**
Paste the following XAML into any resource dictionary:

```xaml
<ControlTheme x:Key="{x:Type Button}" TargetType="Button" BasedOn="{StaticResource {x:Type Button}}"/>
```

The process will crash with `StackOverflowException`. This call loop is repeated:

```
at Avalonia.Markup.Xaml.XamlIl.Runtime.XamlIlRuntimeHelpers+<>c__DisplayClass1_0`1[[System.__Canon, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].<DeferredTransformationFactoryV2>b__0(System.IServiceProvider)
at Avalonia.Controls.ResourceDictionary.TryGetValue(System.Object, System.Object ByRef)
at Avalonia.Controls.ResourceDictionary.TryGetResource(System.Object, System.Object ByRef)
at Avalonia.Markup.Xaml.MarkupExtensions.StaticResourceExtension.ProvideValue(System.IServiceProvider)
at CompiledAvaloniaXaml.!AvaloniaResources+XamlClosure_651.Build(System.IServiceProvider)
```

**Expected behavior**
I expect my `ControlTheme` to extend and replace the base control theme.

**Desktop (please complete the following information):**

- OS: Windows 10
- Version 11 preview 1

### PR DESCRIPTION
**Pull request by maxkatz6**
## How was the solution implemented (if it's not obvious)?

If found resource is the same, as the previous deferred resource, then it will skip current one and continue searching for another resource.
Definitely feels like a hack. Ideally it could detect StaticResource usage with the same key compile time, and inject additional code there, but that would be way overcomplicated. 
Obviously, it won't handle the case with deeper levels of recursion (i.e. Res1 -> Res2 -> Res1), but it's not common either.

## Checklist

- [x] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Fixed issues

Fixes #8831
Fixes #8677

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0032790-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0032794-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

File: src/Avalonia.Base/Controls/ResourceDictionary.cs [hunk 1]
```diff
@@ -15,6 +15,7 @@ namespace Avalonia.Controls
     /// </summary>
     public class ResourceDictionary : IResourceDictionary
     {
+        private DeferredItem? s_lastDeferredItem;
```
Comment (grokys) [inline src/Avalonia.Base/Controls/ResourceDictionary.cs:18]: This has the `s_` prefix but isn't static. Should it be?(I don't think it should?)

File: src/Avalonia.Base/Controls/ResourceDictionary.cs [hunk 2]
```diff
@@ -241,12 +242,27 @@ public bool TryGetValue(object key, out object? value)
             {
                 if (value is DeferredItem deffered)
                 {
-                    _inner[key] = value = deffered.Factory(null) switch
+                    // Avoid simple reentrancy, which could commonly occur on redefining the resource.
+                    if (s_lastDeferredItem == deffered)
```
Comment (grokys) [inline src/Avalonia.Base/Controls/ResourceDictionary.cs:246]: Would it make more sense to check based on the key?

**Review by grokys — CHANGES_REQUESTED** [review]

**Comment by grokys** [comment]
Honestly, I don't think this is too much of a hack.

File: src/Avalonia.Base/Controls/ResourceDictionary.cs (hunk 2 above)
Comment (maxkatz6) [reply src/Avalonia.Base/Controls/ResourceDictionary.cs:246]: Yes, I was thinking about key as an alternative solution

File: src/Avalonia.Base/Controls/ResourceDictionary.cs (hunk 1 above)
Comment (maxkatz6) [reply src/Avalonia.Base/Controls/ResourceDictionary.cs:18]: Yep, it was static, but then I changed. Forgot to adjust the name as well.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0032828-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Review by grokys — APPROVED** [review]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0032941-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Pull request merged**